middleware/tokenauth/tokenauth.go

```diff
@@ -0,0 +1,140 @@
+// Package tokenauth provides jwt token authorisation middleware
+// supports HMAC, RSA, ECDSA, RSAPSS algorithms
+// uses github.com/dgrijalva/jwt-go for jwt implementation
+package tokenauth
+
+import (
+	"github.com/dgrijalva/jwt-go"
+	"github.com/gobuffalo/envy"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strings"
+
+	"github.com/gobuffalo/buffalo"
+	"github.com/pkg/errors"
+)
+
+var (
+	// ErrTokenInvalid is returned when the token provided is invalid
+	ErrTokenInvalid = errors.New("token invalid")
+	// ErrNoToken is returned if no token is supplied in the request.
+	ErrNoToken = errors.New("token not found in request")
+	// ErrBadSigningMethod is returned if the token sign method in the request
+	// does not match the signing method used
+	ErrBadSigningMethod = errors.New("unexpected signing method")
+)
+
+// Options for the JWT middleware
+// SecretKey will be used by HMAC algorithm
+// PublicKeyFile is filename pointing to the location of the public key file
+// for verifying jwt for RSA, ECDSA, RSAPSS algorithms
+type Options struct {
+	SignMethod    jwt.SigningMethod
+	SecretKey     string
+	PublicKeyFile string
+}
+
+// Middleware enables jwt token verification
+// if no Sign method is provided, by default uses HMAC
+func Middleware(options Options) buffalo.MiddlewareFunc {
+	// set sign method to HMAC if not provided
+	if options.SignMethod == nil {
+		options.SignMethod = jwt.SigningMethodHS256
+	}
+	// get key for validation
+	key, err := options.getKey()
+	// if error on getting key exit.
+	if err != nil {
+		log.Fatal(errors.Wrap(err, "couldn't get key"))
+	}
+	return func(next buffalo.Handler) buffalo.Handler {
+		return func(c buffalo.Context) error {
+			// get Authorisation header value
+			authString := c.Request().Header.Get("Authorization")
+
+			tokenString, err := getJwtToken(authString)
+			// if error on getting the token, return with status unauthorized
+			if err != nil {
+				return c.Error(http.StatusUnauthorized, err)
+			}
+
+			// validating and parsing the tokenString
+			token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
+				// Validating if algorithm used for signing is same as the algorithm in token
+				if token.Method.Alg() != options.SignMethod.Alg() {
+					return nil, ErrBadSigningMethod
+				}
+				return key, nil
+			})
+			// if error validating jwt token, return with status unauthorized
+			if err != nil {
+				return c.Error(http.StatusUnauthorized, err)
+			}
+
+			// set the claims as context parameter.
+			// so that the actions can use the claims from jwt token
+			c.Set("claims", token.Claims)
+			// calling next handler
+			err = next(c)
+
+			return err
+		}
+	}
+}
+
+// helper function to get key for validation
+// checks the algorithm used for signing and returns key according to the algorithm used.
+// eg HMAC used a key string and RSA uses public key to validate
+func (options Options) getKey() (interface{}, error) {
```

Comment:
This all seems a bit much, and doesn't allow for users to add their own way of getting a key, such as from a database, or other store.

What do you think of doing something like this:

```go
type Options struct {
	SignMethod jwt.SigningMethod
	GetKey     func(jwt.SigningMethod) (interface{}, error)
}

func GetHMACKey(jwt.SigningMethod) (interface{}, error) {
	key, err := envy.MustGet("JWT_SECRET")
	return []byte(key), err
}

func Middleware(options Options) buffalo.MiddlewareFunc {
	// set sign method to HMAC if not provided
	if options.SignMethod == nil {
		options.SignMethod = jwt.SigningMethodHS256
	}

	if options.GetKey == nil {
		options.GetKey = GetHMACKey
	}

	// ....
}
```

Then we can add a few "default" implementation functions, like the GetHMACKey example. Then people can easily plugin one of the default methods, and if none of them work for their situation, it's easy enough for the to add their own implementations. 

Thoughts?

Reply:
Wow that is neat, i will change the implementation.
instead of setting default `GetKey` to `GetHMACKey` , can we do something like.
```
if options.GetKey == nil {
	options.GetKey = selectGetKeyFunc(options.SignMethod)
}
```
so we get the `GetKey` method that is appropriate to the sign method.
and `GetKey` never uses the signing method so, it can be `GetKey func() (interface{}, error)`